### DISCUSSION TITLE
Attempting to use RateLimitLayer in `tower::limit::rate` on Axum server not working

### DISCUSSION THREAD

**Discussion by khalludi** [tokio-rs/axum#1734] [Q&A] [answered]
Long story short, I'm trying to extend an existing rate limiter. I copied the code exactly and made only one change in `rate_limit/service.rs` [lines 21-33](https://github.com/khalludi/token-bucket-rl/blob/3631caf9a164f7982a2ae89d84d0e8d5d222c18d/src/rate_limit/service.rs#L21) where I add a clone implementation to `RateLimit`. The change was needed because all layers need to be clonable to add to the Axum router's `layer()` func (see [`main.rs`](https://github.com/khalludi/token-bucket-rl/blob/3631caf9a164f7982a2ae89d84d0e8d5d222c18d/src/main.rs#L31)). 

The problem is, when I try to modify the struct's field in the Service's `call()` method ([see here](https://github.com/khalludi/token-bucket-rl/blob/3631caf9a164f7982a2ae89d84d0e8d5d222c18d/src/rate_limit/service.rs#L123)), the state is not updating. In the output below, you can see that the local state inside the `call()` method is updated ('New Until' tag), however the global state inside the struct is not updated (same 'Until' time tag on each request).

My question is, what should I be doing so that I can modify the state field (or any field) within the `call()` method of a service? Alternatively, would creating a similar layer in Axum be the correct way for developing layers?

Output:
```
Now: Instant { t: 414379.3779662s }
Until: Instant { t: 414376.774569s }
Rem: 5
New Until: Instant { t: 414439.3779662s }
-------
Now: Instant { t: 414381.2349382s }
Until: Instant { t: 414376.774569s }
Rem: 5
New Until: Instant { t: 414441.2349382s }
-------
Now: Instant { t: 414382.1941313s }
Until: Instant { t: 414376.774569s }
Rem: 5
New Until: Instant { t: 414442.1941313s }
-------
Now: Instant { t: 414383.3867334s }
Until: Instant { t: 414376.774569s }
Rem: 5
New Until: Instant { t: 414443.3867334s }
```

Github reproducible code: https://github.com/khalludi/token-bucket-rl

**Comment by davidpdrsn** [comment, marked as answer]
There is a reason tower's rate limiting middleware doesn't implement `Clone`. By doing what you're doing your essentially bypassing the limiting by cloning everything. The clones don't actually share the underlying service, or the bookkeeping for how many requests have been sent.

I answered a similar question previously. See https://github.com/tokio-rs/axum/discussions/987

**Reply by psionic-k** [reply]
Seems like the better approach is to implement clone on an outer struct that hides the thread safety of an inner rate limit implementation.
